[quant][pt2][be] Rewrite QAT annotations using subgraph matcher

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113714
* __->__ #113709

Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D51366525](https://our.internmc.facebook.com/intern/diff/D51366525)

andrewor14 committed:
[quant][pt2][be] Rewrite QAT annotations using subgraph matcher

Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -318,129 +325,114 @@ def _annotate_conv_relu(
+        def _conv_bn_relu(x, conv_weight, conv_bias, bn_weight, bn_bias, bn_rm, bn_rv):
+            conv = conv_fn(x, conv_weight, conv_bias)
+            bn = F.batch_norm(conv, bn_rm, bn_rv, bn_weight, bn_bias, training=True)
+            relu = F.relu(bn)
```
nit: is it possible to also merge this with the other _get_pattern function as well? e.g. `_get_pattern(conv_fn: Callable, has_relu: bool)`?

andrewor14 replied:
Interesting. Yeah I can try

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -318,129 +325,114 @@ def _annotate_conv_relu(
+    for match in matches:
+        name_node_map = match.name_node_map
+        input_node = name_node_map["input"]
+        conv_node = name_node_map["conv"]
```
nit: to be more robust to decompositions, we'd like to get first user of input_node, first user of weight_node and first user of bias_node etc. and annotate these nodes instead

andrewor14 replied:
Ok, but isn't that the whole point of `name_node_map`, so that the user doesn't have to manually walk through the graph?

jerryzh168 replied:
the goal is for easier access of nodes, and in this case we want access to input and output of the pattern only, since only these are robust

andrewor14 replied:
Ok then maybe we should update the tutorials :): https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -318,130 +325,139 @@ def _annotate_conv_relu(
+        weight_user = list(weight_node.users.keys())[0]
+        if weight_user is not input_user:
+            raise ValueError("Expected weight user to be the same as input user")
+        conv_node = input_user
```
This still assumes conv_node is not decomposed. (when it's decomposed, input_user, weight_user and bias_user could be different nodes, e.g. linear is a good example)

I think you'll need to annotate `input_user`, `weight_user` and `bias_user` separately to be 100% robust to decompositions of the conv op

andrewor14 replied:
As discussed offline, we will just use the `conv_node` returned from the subgraph matcher here and just validate the conv args for now. This is because we can't always rely on the first user of the input node, for example in resnet18 where there are skip connections

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51366525).

andrewor14 committed:
Update on "[quant][pt2][be] Rewrite QAT annotations using subgraph matcher"


Summary: This is the recommended way to write quantizers according
to https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow.
It is agnostic to changes in the aten IR and can be easily extended
to support conv1d-bn and conv3d-bn fusion patterns in the future.
This is the first step towards rewriting XNNPACKQuantizer using
this subgraph matcher.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D51366525](https://our.internmc.facebook.com/intern/diff/D51366525)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51366525).

jerryzh168 reviewed (approved):
looks good, thanks for addressing the comments!

andrewor14 commented:
@pytorchbot merge
